ocaml/setup-ocaml
# setup-ocaml v2

This was actually all rewritten from scratch, and in the process many changes were made to make things better and better. (Mainly for Windows and Build Performance.)

### Now the action does the following:

1. Change the file system behavioral parameters
   - **Windows only**
1. Retrieve the Cygwin cache
   - **Windows only**
   - If the cache already exists
1. Retrieve the opam cache
   - If the cache already exists
1. Prepare the Cygwin environment
   - **Windows only**
1. Save the Cygwin cache
   - **Windows only**
1. Install opam
1. Initialise the opam state
1. Install the OCaml compiler
   - If the opam cache was not hit
1. Save the opam cache
   - If the opam cache was not hit
1. Retrieve the opam download cache
1. Install depext
   - On Windows, not only `opam-depext` is installed, but `depext-cygwinports` is installed as well
1. Retrieve the dune cache
   - If the dune cache feature is enabled
   - If the cache already exists
1. Install the latest dune and enable the dune cache feature
    - If the dune cache feature is enabled
1. Pin the opam files, if they exist
    - If the pin feature is not disabled
    - If there is an opam file in the workspace that matches the glob pattern
1. Install the system dependencies required by the opam files via depext
    - If the depext feature is enabled
    - If there is an opam file in the workspace that matches the glob pattern

Post:

The reason for not caching opam stuff in the post stage (more precisely, why you can't) is due to the size of the cache and repeatability. They should be cached immediately after initialisation to minimize the size of the cache.

1. Remove oldest dune cache files to free space
    - If the dune cache feature is enabled
1. Save the dune cache
    - If the dune cache feature is enabled
1. Save the opam download cache

### Inputs

- `ocaml-compiler`: The OCaml compiler packages to initialise. The packages must be separated by the comma.
  - required: **Yes**
- `opam-repository`: The URL of the repository to fetch the packages from.
  - required: **No**
  - default:
- `opam-pin`: Enable the automation feature for opam pin.
  - required: **No**
  - default: `true`
- `opam-depext`: Enable the automation feature for opam depext.
  - required: **No**
  - default: `true`
- `opam-depext-flags`: The flags for the opam depext command. The flags must be separated by the comma.
  - required: **No**
  - default:
- `opam-local-packages`: The local packages to be used by `opam-pin` or `opam-depext`. See [`@actions/glob`](https://github.com/actions/toolkit/tree/main/packages/glob) for supported patterns.
  - required: **No**
  - default: `*.opam`
- `opam-disable-sandboxing`: Disable the opam sandboxing feature.
  - required: **No**
  - default: `false`
- `dune-cache`: Enable the dune cache feature.
  - required: **No**
  - default: `false`
- `cache-prefix`: The prefix of the cache keys.
  - required: **No**
  - default: `v1`
- `github-token`: **DO NOT SET THIS.** The API token to be used by the action to avoid rate-limiting when calling the GitHub API internally.
  - required: **No**
  - default: `${{ github.token }}`

#### Follow-up to misunderstandable parts:

- `opam-disable-sandboxing` is always meaningless on Windows.
- `opam-local-packages` actually allows complex glob patterns that span multiple lines:

  See [`@actions/glob`](https://github.com/actions/toolkit/tree/main/packages/glob) for supported patterns.

  ```yml
  - name: Use OCaml ${{ matrix.ocaml-compiler }}
    uses: avsm/setup-ocaml@v2
    with:
      ocaml-compiler: ${{ matrix.ocaml-compiler }}
      opam-local-packages: |
        *.opam
        !exclude.opam
  ```
- In what kind of cases should I use `cache-prefix`?
  - In a use case like https://github.com/avsm/setup-ocaml/issues/6, the cache is often not managed as expected because the action only generates the cache key based on the platform and architecture.
- Do I need to create an API token, set it as a repository secret, and pass it to the action via `github-token`?!
  - No. This is a hidden hack that allows the action to use the API token automatically generated each time in the action runner, which has minimal privileges. You don't have to set it up by yourself.

### Comparison

#### Set-up

Here is a simple benchmark. However, this is not really a fair benchmark. In v2, if the cache wasn't hit, the time spent creating the cache and uploading it to the server is included, so they are a bit longer than the actual set-up time.
In addition, the set-up time on macOS is very unstable. The Actions team is currently working to fix it, but due to an issue such as https://github.com/actions/virtual-environments/issues/2707, the set-up time for each run is completely unreliable.

|                    | macos-latest | ubuntu-latest | windows-latest |
| ------------------ | ------------ | ------------- | -------------- |
| v1                 | 9m 0s        | 4m 30s        | 3m 20s         |
| v2 (with cache)    | 0m 50s       | 0m 30s        | 1m 45s         |
| v2 (without cache) | 4m 55s       | 4m 40s        | 3m 5s          |

#### Others

Here is a result of benchmarking several times on a medium-sized repository using ubuntu-latest VM.

|                      | `opam install . --deps-only --with-test --with-doc` | `opam exec -- make build` | `opam exec -- make test` |
| -------------------- | --------------------------------------------------- | ------------------------- | ------------------------ |
| v1                   | 8m 25s                                              | 0m 18s                    | 0m 6s                    |
| v2 (with dune cache) | 5m 10s                                              | 0m 4s                     | 0m 1s                    |

### Relevant issues and pull requests

#### General

##### Issues

- https://github.com/avsm/setup-ocaml/issues/1
- https://github.com/avsm/setup-ocaml/issues/2
- https://github.com/avsm/setup-ocaml/issues/6
- https://github.com/avsm/setup-ocaml/issues/36
- https://github.com/avsm/setup-ocaml/issues/65
- https://github.com/avsm/setup-ocaml/issues/67
- https://github.com/avsm/setup-ocaml/issues/68
- https://github.com/avsm/setup-ocaml/issues/73

##### Pull requests

- [x] https://github.com/actions/toolkit/pull/609: The PATH hacks on Windows are needed until this PR is merged.
- [ ] https://github.com/actions/toolkit/pull/610: Same as above.

#### Cache

##### Issues

- https://github.com/actions/virtual-environments/issues/2858

#### Dune cache

##### Issues

- https://github.com/ocaml/dune/issues/3987
- https://github.com/ocaml/dune/issues/4166
- https://github.com/ocaml/dune/issues/4167

##### Pull requests

- [ ] https://github.com/ocaml/opam/pull/4535: The dune cache feature is not supported on macOS until opam 2.0.9 is released. (See: https://github.com/ocaml/opam/pull/4547)
- [x] https://github.com/ocaml/dune/pull/4406: The dune cache feature is not supported on Windows until dune 2.8.5 is released. (See: https://github.com/ocaml/opam-repository/pull/18405)

**Acknowledgement: This work was done under great supervision of @dra27.**
